module2/gift-certificate-system/web/src/main/java/com/epam/esm/controller/TagController.java

```diff
@@ -9,38 +12,95 @@
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
+import org.springframework.validation.BeanPropertyBindingResult;
+import org.springframework.validation.BindingResult;
 import org.springframework.web.bind.annotation.*;
+import org.springframework.web.util.UriComponentsBuilder;
 
+import javax.validation.Valid;
+import java.util.ArrayList;
 import java.util.List;
 
 @RestController
 @RequestMapping("/gift-certificates")
 public class TagController {
 
     private final TagService tagService;
+    private final GiftCertificateService certificateService;
     private final TagValidator tagValidator;
 
-    public TagController(TagService tagService, TagValidator tagValidator) {
+    public TagController(TagService tagService, GiftCertificateService certificateService, TagValidator tagValidator) {
         this.tagService = tagService;
+        this.certificateService = certificateService;
         this.tagValidator = tagValidator;
     }
 
-    @PatchMapping(path = "/{id}", consumes = "application/json-patch+json")
-    public ResponseEntity<Integer> updateGiftCertificate(@PathVariable Long id, @RequestBody JsonPatch patch) {
+    @GetMapping(value = "/tags")
+    public List<Tag> findAllTags() {
+        return tagService.findAllTags();
+    }
+
+    @GetMapping(value = "tags/{id}")
+    public Tag findTagById(@PathVariable Long id) {
+        return tagService.findTagById(id);
+    }
+
+    @PostMapping(path = "/tags", consumes = "application/json", produces = "application/json")
+    public ResponseEntity<Tag> addTag(@RequestBody @Valid Tag tag,
+                                                              UriComponentsBuilder ucBuilder) {
+
+        BindingResult result = new BeanPropertyBindingResult(tag, "tag");
+        tagValidator.validate(tag, result);
+        if (result.hasErrors()) {
+            return new ResponseEntity(result.getAllErrors(), HttpStatus.BAD_REQUEST);
+        } else {
+            Long tagId = tagService.saveTag(tag);
+            HttpHeaders headers = new HttpHeaders();
+            headers.setLocation(ucBuilder.path("/tags/{id}").buildAndExpand(tagId).toUri());
+            return new ResponseEntity<>(headers, HttpStatus.CREATED);
+        }
+    }
+
+    @PatchMapping(path = "tags/{id}", consumes = "application/json-patch+json")
+    public ResponseEntity<Tag> updateTag(@PathVariable Long id,
+                                                         @RequestBody JsonPatch patch,
+                                                         UriComponentsBuilder ucBuilder) {
         try {
             Tag oldTag = tagService.findTagById(id);
             Tag tagPatched = applyPatchToTag(patch, oldTag);
-            tagService.updateTag(tagPatched);
-            return ResponseEntity.ok().build();
+            BindingResult result = new BeanPropertyBindingResult(tagPatched, "tagPatched");
+            tagValidator.validate(tagPatched, result);
```

Comment:
needs to encapsulate all logic from controller to service method:
and it will be like:

var ... = tagService.updateTag(...);